# Data Analytics Exercises 6 Models

### XXXX XXXX 



### Problem 1. Wines


[Here](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p01_wine.csv) is some data on Portuguese wines. 

Drop rows with missing values.

Use logistic regression to predict the type (white/red) from the other fields.

Split train/test set 70/30 %. Print the score and the confusion matrix.


In [17]:
import pandas as pd

df = pd.read_csv('data/exrc06p01_wine.csv')
df=df.dropna()
df.head()

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6463 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   type                  6463 non-null   object 
 1   fixed acidity         6463 non-null   float64
 2   volatile acidity      6463 non-null   float64
 3   citric acid           6463 non-null   float64
 4   residual sugar        6463 non-null   float64
 5   chlorides             6463 non-null   float64
 6   free sulfur dioxide   6463 non-null   float64
 7   total sulfur dioxide  6463 non-null   float64
 8   density               6463 non-null   float64
 9   pH                    6463 non-null   float64
 10  sulphates             6463 non-null   float64
 11  alcohol               6463 non-null   float64
 12  quality               6463 non-null   int64  
dtypes: float64(11), int64(1), object(1)
memory usage: 706.9+ KB


In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler


# the logistic regression model
classifier = LogisticRegression()

X = df.drop('type', axis=1)
y = df['type']

# scaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [20]:

from sklearn.model_selection import train_test_split

# splitting to train and test sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3,  random_state=42)

classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)


print('score (scaled fit):', classifier.score(X_test, y_test))

print(y_pred)


score (scaled fit): 0.9948427024239299
['white' 'white' 'red' ... 'red' 'white' 'red']


In [21]:

from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix


array([[ 472,    5],
       [   5, 1457]])

In [22]:
#The same with pipeline


import pandas as pd
from sklearn.model_selection import train_test_split

from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
#from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
my_classifier = make_pipeline(
    StandardScaler(),
    LogisticRegression()
)
# my_classifier # classify 
# override the pipeline above to ignore the scaling part...

myscaler = StandardScaler()
# ... and do the scaling separately

df = pd.read_csv(
    'data/exrc06p01_wine.csv'
)
df=df.dropna()
X = df.drop('type', axis=1)
y = df['type']

myscaler.fit(X)
# same as StandardScaler().fit()
# myscaler now contains the info about how
# the columns of X need to be scaled to normality

X_scaled = pd.DataFrame(
    myscaler.transform(X), # the actual scaling happens here
    columns=X.columns
)
# a mere myscaler.transform(X) would produce an array
# but that's minor... made a dataframe in order to compare:

# print(X)
# print(X_scaled)
# comment out the above to see the unscaled and scaled data

# then fit with scaling
# note: X_scaled used below
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3,  random_state=42)
my_classifier.fit(X_train, y_train)

y_pred = my_classifier.predict(X_test) #added line

print(
    'score (scaled fit):',
    my_classifier.score(X_test, y_test)
)


score (scaled fit): 0.9948427024239299


In [23]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix


array([[ 472,    5],
       [   5, 1457]])

In [24]:
print(y_pred)

['white' 'white' 'red' ... 'red' 'white' 'red']


### Problem 2. Voices
[Here](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p02_voice.csv) is some data on human voices ([column info](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p02_voice.txt)).
 
Predict the label from the other fields using a support vector machine.

Split train/test set 70/30 %.

Print the score and the confusion matrix.


In [25]:
import pandas as pd

df = pd.read_csv('data/exrc06p02_voice.csv')
df=df.dropna()
df.head()

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0  0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1  0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2  0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3  0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4  0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   

          kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0   274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1   634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2  1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3     4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4     4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   

     maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0  0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1  0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2  0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3  0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4  0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  

[5 rows x 21 columns]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3168 entries, 0 to 3167
Data columns (total 21 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   meanfreq  3168 non-null   float64
 1   sd        3168 non-null   float64
 2   median    3168 non-null   float64
 3   Q25       3168 non-null   float64
 4   Q75       3168 non-null   float64
 5   IQR       3168 non-null   float64
 6   skew      3168 non-null   float64
 7   kurt      3168 non-null   float64
 8   sp.ent    3168 non-null   float64
 9   sfm       3168 non-null   float64
 10  mode      3168 non-null   float64
 11  centroid  3168 non-null   float64
 12  meanfun   3168 non-null   float64
 13  minfun    3168 non-null   float64
 14  maxfun    3168 non-null   float64
 15  meandom   3168 non-null   float64
 16  mindom    3168 non-null   float64
 17  maxdom    3168 non-null   float64
 18  dfrange   3168 non-null   float64
 19  modindx   3168 non-null   float64
 20  label     3168 non-null   obje

In [27]:
from sklearn.model_selection import train_test_split

from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC


X = df.drop('label', axis=1)
y = df['label']

my_classifier = SVC()


scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

    
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3,  random_state=42)

my_classifier.fit(X_train, y_train)
y_pred = my_classifier.predict(X_test)

print('score (scaled fit):', my_classifier.score(X_test, y_test))
print(y_pred[:50])

score (scaled fit): 0.9810725552050473
['female' 'male' 'male' 'female' 'female' 'female' 'male' 'male' 'female'
 'male' 'female' 'female' 'female' 'female' 'female' 'female' 'female'
 'female' 'female' 'male' 'male' 'female' 'male' 'female' 'male' 'female'
 'male' 'male' 'male' 'male' 'male' 'female' 'female' 'male' 'male'
 'female' 'male' 'male' 'male' 'male' 'female' 'male' 'male' 'female'
 'female' 'male' 'male' 'female' 'male' 'male']


In [28]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[445,   7],
       [ 11, 488]])

In [30]:
# The same with pipeline and without random state

import pandas as pd
from sklearn.model_selection import train_test_split

from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC

my_classifier = make_pipeline(
    StandardScaler(),
    SVC()
)
# my_classifier = SVC() # classify with support vector
# override the pipeline above to ignore the scaling part...

myscaler = StandardScaler()
# ... and do the scaling separately

df = pd.read_csv(
    'data/exrc06p02_voice.csv'
)

df=df.dropna()

X = df.drop('label', axis=1)
y = df['label']

myscaler.fit(X)
# same as StandardScaler().fit()
# myscaler now contains the info about how
# the columns of X need to be scaled to normality

X_scaled = pd.DataFrame(
    myscaler.transform(X), # the actual scaling happens here
    columns=X.columns
)
# a mere myscaler.transform(X) would produce an array
# but that's minor... made a dataframe in order to compare:

# print(X)
# print(X_scaled)
# comment out the above to see the unscaled and scaled data

# first fit without scaling
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.3)
my_classifier.fit(X_train, y_train)
y_pred = my_classifier.predict(X_test)

print(
    'score (unscaled fit):',
    my_classifier.score(X_test, y_test)
)

# then fit with scaling
# note: X_scaled used below
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y,test_size=0.3)
my_classifier.fit(X_train, y_train)
y_pred = my_classifier.predict(X_test)

print(
    'score (scaled fit):',
    my_classifier.score(X_test, y_test)
)


score (unscaled fit): 0.9800210304942166
score (scaled fit): 0.9810725552050473


In [31]:
print(y_pred[:50])



['female' 'female' 'female' 'female' 'male' 'female' 'female' 'female'
 'male' 'male' 'female' 'male' 'female' 'male' 'female' 'female' 'female'
 'male' 'male' 'female' 'male' 'male' 'male' 'female' 'female' 'female'
 'male' 'female' 'male' 'male' 'male' 'male' 'male' 'male' 'female' 'male'
 'female' 'male' 'female' 'female' 'female' 'female' 'female' 'male'
 'male' 'male' 'male' 'male' 'female' 'female']


In [32]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[464,  11],
       [  7, 469]])

### Problem 3. NBA
[Here](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p03_nba.csv) is some data on NBA basketball players in their first season ([column info](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p03_nba.csv)).

Last column tells if player's career has exceed 5 years or not.

Fill missing values with the field median.

Try to predict if the career has exceeded 5 years or not by using both logistic regression and a support vector machine. Print scores and confusion matrices. Split train/test data as you wish. Compare the results.


In [34]:
import pandas as pd

df = pd.read_csv('data/exrc06p03_nba.csv')

df.head()

Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   

   FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8          1.0  

[5 rows x 21 columns]

In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1340 entries, 0 to 1339
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         1340 non-null   object 
 1   GP           1340 non-null   int64  
 2   MIN          1340 non-null   float64
 3   PTS          1340 non-null   float64
 4   FGM          1340 non-null   float64
 5   FGA          1340 non-null   float64
 6   FG%          1340 non-null   float64
 7   3P Made      1340 non-null   float64
 8   3PA          1340 non-null   float64
 9   3P%          1329 non-null   float64
 10  FTM          1340 non-null   float64
 11  FTA          1340 non-null   float64
 12  FT%          1340 non-null   float64
 13  OREB         1340 non-null   float64
 14  DREB         1340 non-null   float64
 15  REB          1340 non-null   float64
 16  AST          1340 non-null   float64
 17  STL          1340 non-null   float64
 18  BLK          1340 non-null   float64
 19  TOV   

In [36]:
df.describe()

GP          MIN          PTS          FGM          FGA  \
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000   
mean     60.414179    17.624627     6.801493     2.629104     5.885299   
std      17.433992     8.307964     4.357545     1.683555     3.593488   
min      11.000000     3.100000     0.700000     0.300000     0.800000   
25%      47.000000    10.875000     3.700000     1.400000     3.300000   
50%      63.000000    16.100000     5.550000     2.100000     4.800000   
75%      77.000000    22.900000     8.800000     3.400000     7.500000   
max      82.000000    40.900000    28.200000    10.200000    19.800000   

               FG%      3P Made          3PA          3P%          FTM  \
count  1340.000000  1340.000000  1340.000000  1329.000000  1340.000000   
mean     44.169403     0.247612     0.779179    19.308126     1.297687   
std       6.137679     0.383688     1.061847    16.022916     0.987246   
min      23.800000     0.000000     0.000000     0.000000     0.000000   
25%      40.200000     0.000000     0.000000     0.000000     0.600000   
50%      44.100000     0.100000     0.300000    22.400000     1.000000   
75%      47.900000     0.400000     1.200000    32.500000     1.600000   
max      73.700000     2.300000     6.500000   100.000000     7.700000   

               FTA          FT%         OREB         DREB          REB  \
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000   
mean      1.821940    70.300299     1.009403     2.025746     3.034478   
std       1.322984    10.578479     0.777119     1.360008     2.057774   
min       0.000000     0.000000     0.000000     0.200000     0.300000   
25%       0.900000    64.700000     0.400000     1.000000     1.500000   
50%       1.500000    71.250000     0.800000     1.700000     2.500000   
75%       2.300000    77.600000     1.400000     2.600000     4.000000   
max      10.200000   100.000000     5.300000     9.600000    13.900000   

               AST          STL          BLK          TOV  TARGET_5Yrs  
count  1340.000000  1340.000000  1340.000000  1340.000000  1340.000000  
mean      1.550522     0.618507     0.368582     1.193582     0.620149  
std       1.471169     0.409759     0.429049     0.722541     0.485531  
min       0.000000     0.000000     0.000000     0.100000     0.000000  
25%       0.600000     0.300000     0.100000     0.700000     0.000000  
50%       1.100000     0.500000     0.200000     1.000000     1.000000  
75%       2.000000     0.800000     0.500000     1.500000     1.000000  
max      10.600000     2.500000     3.900000     4.400000     1.000000

In [37]:
#filling missing values with median

num_col = df.iloc[:, 1:]

df.iloc[:,1:]=num_col.fillna(num_col.median())

df.head()


Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0   Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1  Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2   JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3      Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4      Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   

   FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0  2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1  3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2  1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3  1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4  1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8          1.0  

[5 rows x 21 columns]

In [38]:
#logistic regression

from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
import numpy as np

# the logistic regression model
lr_classifier = LogisticRegression()

X = df.drop('TARGET_5Yrs', axis=1)
y = df['TARGET_5Yrs']


# scaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X.select_dtypes(np.number))



In [39]:

from sklearn.model_selection import train_test_split

# splitting to train and test sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3,  random_state=42)

lr_classifier.fit(X_train, y_train)
y_pred = lr_classifier.predict(X_test)


print('score (scaled fit):', lr_classifier.score(X_test, y_test))

print(y_pred)

score (scaled fit): 0.7213930348258707
[0. 0. 1. 0. 1. 1. 1. 0. 0. 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 0. 0. 1. 1.
 0. 1. 0. 1. 0. 0. 1. 1. 0. 0. 1. 0. 1. 1. 1. 1. 1. 0. 1. 1. 1. 0. 0. 0.
 0. 1. 1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1. 1. 0. 1. 0.
 1. 1. 1. 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 0. 1. 1. 0. 1. 1. 0. 0. 1. 1.
 1. 1. 1. 0. 1. 1. 1. 1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 1. 1. 1. 1. 1. 0. 0.
 1. 1. 1. 1. 1. 1. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1.
 1. 1. 0. 1. 1. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 0. 1.
 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 0. 1. 0. 1. 1. 1. 0. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 0. 0. 0. 1. 1. 1. 1. 0. 0. 1. 1. 1. 1. 1. 0. 1. 1. 0.
 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 0. 0. 0.
 0. 1. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1.
 1. 0. 1. 1.

In [40]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[ 78,  78],
       [ 34, 212]])

In [41]:
#svm

from sklearn.svm import SVC


svm_classifier = SVC()

svm_classifier.fit(X_train, y_train)

y_pred = svm_classifier.predict(X_test)

print('score (scaled fit):', svm_classifier.score(X_test, y_test))
print(y_pred)

score (scaled fit): 0.7213930348258707
[0. 0. 1. 0. 0. 1. 1. 0. 0. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 0. 0. 1.
 0. 1. 0. 1. 0. 0. 1. 1. 0. 0. 1. 0. 1. 1. 1. 0. 1. 0. 1. 1. 1. 0. 0. 0.
 0. 1. 1. 1. 0. 1. 0. 1. 0. 0. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1. 1. 0. 1. 0.
 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1. 0. 0. 1. 1.
 1. 1. 1. 0. 1. 1. 0. 1. 1. 0. 1. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0. 0.
 1. 1. 0. 1. 1. 1. 0. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0. 1. 1.
 1. 1. 1. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 0. 1.
 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1.
 1. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 0.
 1. 0. 0. 1. 1. 0. 1. 1. 1. 1. 1. 1. 0. 1. 0. 1. 0. 1. 1. 1. 0. 1. 1. 1.
 1. 1. 1. 1. 1. 1. 0. 0. 0. 1. 1. 1. 1. 0. 0. 1. 1. 1. 1. 1. 0. 1. 1. 0.
 1. 1. 1. 1. 1. 0. 1. 1. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 0. 0.
 1. 1. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 1. 1. 1. 0. 1. 0. 1. 1. 1. 1. 1. 1.
 1. 0. 1. 1.

In [43]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[ 80,  76],
       [ 36, 210]])

The SVM model's accuracy is similar to that of the logistic regression model. However, there is some difference in the confusion matrix.
SVM has slightly better true negatives (80) and the false positives of SVM model (76) were a bit less than for the logistic regression model. 
However, the false negatives increased for SVM (36) and true positives decreased with SVM to 210. So, we cannot say clearly, which model is better.

### Problem 4.  Mushrooms
[Here](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p04_mushrooms.csv) is some data on mushrooms ([column info](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p04_mushrooms.txt)).

Try to predict the class (edible or poisonous) from the other fields. Use whatever you want!

Fields are categorial so one-hot-encoding is needed.


In [44]:
import pandas as pd

df = pd.read_csv('data/exrc06p04_mushrooms.csv')
df.head(15)

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0      p         x           s         n       t    p               f   
1      e         x           s         y       t    a               f   
2      e         b           s         w       t    l               f   
3      p         x           y         w       t    p               f   
4      e         x           s         g       f    n               f   
5      e         x           y         y       t    a               f   
6      e         b           s         w       t    a               f   
7      e         b           y         w       t    l               f   
8      p         x           y         w       t    p               f   
9      e         b           s         y       t    a               f   
10     e         x           y         y       t    l               f   
11     e         x           y         y       t    a               f   
12     e         b           s         y       t    a               f   
13     p         x           y         w       t    p               f   
14     e         x           f         n       f    n               f   

   gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0             c         n          k  ...                        s   
1             c         b          k  ...                        s   
2             c         b          n  ...                        s   
3             c         n          n  ...                        s   
4             w         b          k  ...                        s   
5             c         b          n  ...                        s   
6             c         b          g  ...                        s   
7             c         b          n  ...                        s   
8             c         n          p  ...                        s   
9             c         b          g  ...                        s   
10            c         b          g  ...                        s   
11            c         b          n  ...                        s   
12            c         b          w  ...                        s   
13            c         n          k  ...                        s   
14            w         b          n  ...                        f   

   stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                       w                      w         p          w   
1                       w                      w         p          w   
2                       w                      w         p          w   
3                       w                      w         p          w   
4                       w                      w         p          w   
5                       w                      w         p          w   
6                       w                      w         p          w   
7                       w                      w         p          w   
8                       w                      w         p          w   
9                       w                      w         p          w   
10                      w                      w         p          w   
11                      w                      w         p          w   
12                      w                      w         p          w   
13                      w                      w         p          w   
14                      w                      w         p          w   

   ring-number ring-type spore-print-color population habitat  
0            o         p                 k          s       u  
1            o         p                 n          n       g  
2            o         p                 n          n       m  
3            o         p                 k          s       u  
4            o         e                 n          a       g  
5            o         p                 k          n       g  
6            o         p                 k          n       m  
7            o         p        

In [45]:
# unique values in each columns
for x in df.columns:
    print(x ,':', len(df[x].unique()))

class : 2
cap-shape : 6
cap-surface : 4
cap-color : 10
bruises : 2
odor : 9
gill-attachment : 2
gill-spacing : 2
gill-size : 2
gill-color : 12
stalk-shape : 2
stalk-root : 5
stalk-surface-above-ring : 4
stalk-surface-below-ring : 4
stalk-color-above-ring : 9
stalk-color-below-ring : 9
veil-type : 1
veil-color : 4
ring-number : 3
ring-type : 5
spore-print-color : 9
population : 6
habitat : 7


In [46]:
# the example applied from https://machinelearningmastery.com/one-hot-encoding-for-categorical-data/

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder


X = df.drop('class', axis=1)
y = df['class']


# split the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# one-hot encode input variables
onehot_encoder = OneHotEncoder()

onehot_encoder.fit(X_train)

X_train = onehot_encoder.transform(X_train)
X_test = onehot_encoder.transform(X_test)

# encode target variable
#label_encoder = LabelEncoder() # this is optional to encode also y

#label_encoder.fit(y_train)

#y_train = label_encoder.transform(y_train)
#y_test = label_encoder.transform(y_test)

print(X_train)



  (0, 5)	1.0
  (0, 8)	1.0
  (0, 10)	1.0
  (0, 21)	1.0
  (0, 24)	1.0
  (0, 32)	1.0
  (0, 33)	1.0
  (0, 35)	1.0
  (0, 40)	1.0
  (0, 50)	1.0
  (0, 52)	1.0
  (0, 58)	1.0
  (0, 60)	1.0
  (0, 71)	1.0
  (0, 80)	1.0
  (0, 82)	1.0
  (0, 85)	1.0
  (0, 88)	1.0
  (0, 94)	1.0
  (0, 96)	1.0
  (0, 108)	1.0
  (0, 115)	1.0
  (1, 5)	1.0
  (1, 6)	1.0
  (1, 13)	1.0
  :	:
  (5684, 102)	1.0
  (5684, 108)	1.0
  (5684, 114)	1.0
  (5685, 3)	1.0
  (5685, 6)	1.0
  (5685, 13)	1.0
  (5685, 20)	1.0
  (5685, 27)	1.0
  (5685, 32)	1.0
  (5685, 34)	1.0
  (5685, 35)	1.0
  (5685, 39)	1.0
  (5685, 49)	1.0
  (5685, 51)	1.0
  (5685, 58)	1.0
  (5685, 61)	1.0
  (5685, 71)	1.0
  (5685, 80)	1.0
  (5685, 82)	1.0
  (5685, 85)	1.0
  (5685, 89)	1.0
  (5685, 94)	1.0
  (5685, 102)	1.0
  (5685, 106)	1.0
  (5685, 111)	1.0


In [47]:

# define the model
model = LogisticRegression()

# fit on the training set
model.fit(X_train, y_train)

# predict on test set
y_pred = model.predict(X_test)



print('score:', model.score(X_test, y_test))

print(y_pred)

score: 1.0
['e' 'p' 'p' ... 'p' 'e' 'e']


In [49]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[1257,    0],
       [   0, 1181]])

The model has 100% accuracy.

### Problem 5. Loan status
[Here](https://student.labranet.jamk.fi/~varpha/data_analytics/exrc06p05_loan.txt) is some data on loanees. Last column Loan_Status Y/N, should be predicted from the other field. Use whatever you want.  


Do modifications:
* Categorial fields to numeric (2-value fields to 0/1, multivalue as dummies)
* replace missing values with median
* remove rows with outliers: ApplicantIncome, CoapplicantIncome or LoanAmount over 3 standard deviations away from field average


Check what would be model's probability to Loan_status = Yes with values:

```
Gender                   Male
Married                    No
Dependents                  0
Education            Graduate
Self_Employed              No
ApplicantIncome          2400
CoapplicantIncome        2000
LoanAmount                 36
Loan_Amount_Term          360
Credit_History              1
Property_Area           Urban
```

In [50]:
import pandas as pd

df = pd.read_csv('data/loans.txt')

df.set_index('Loan_ID', inplace=True)

df.to_csv("exrc06p05_loans.csv", index = True)

df.head()

Gender Married  Dependents     Education Self_Employed  \
Loan_ID                                                           
LP001002   Male      No         0.0      Graduate            No   
LP001003   Male     Yes         1.0      Graduate            No   
LP001005   Male     Yes         0.0      Graduate           Yes   
LP001006   Male     Yes         0.0  Not Graduate            No   
LP001008   Male      No         0.0      Graduate            No   

          ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
Loan_ID                                                                      
LP001002             5849                0.0         NaN             360.0   
LP001003             4583             1508.0       128.0             360.0   
LP001005             3000                0.0        66.0             360.0   
LP001006             2583             2358.0       120.0             360.0   
LP001008             6000                0.0       141.0             360.0   

          Credit_History Property_Area Loan_Status  
Loan_ID                                             
LP001002             1.0         Urban           Y  
LP001003             1.0         Rural           N  
LP001005             1.0         Urban           Y  
LP001006             1.0         Urban           Y  
LP001008             1.0         Urban           Y

In [2]:
# dtype of each column 
result = df.dtypes 
 
print(result) 


Gender                object
Married               object
Dependents           float64
Education             object
Self_Employed         object
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Property_Area         object
Loan_Status           object
dtype: object


In [51]:
import numpy as np

num_cols = df.select_dtypes(include=np.number).columns

df[num_cols] = df[num_cols].fillna(df[num_cols].median())

df.head()



Gender Married  Dependents     Education Self_Employed  \
Loan_ID                                                           
LP001002   Male      No         0.0      Graduate            No   
LP001003   Male     Yes         1.0      Graduate            No   
LP001005   Male     Yes         0.0      Graduate           Yes   
LP001006   Male     Yes         0.0  Not Graduate            No   
LP001008   Male      No         0.0      Graduate            No   

          ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
Loan_ID                                                                      
LP001002             5849                0.0       126.0             360.0   
LP001003             4583             1508.0       128.0             360.0   
LP001005             3000                0.0        66.0             360.0   
LP001006             2583             2358.0       120.0             360.0   
LP001008             6000                0.0       141.0             360.0   

          Credit_History Property_Area Loan_Status  
Loan_ID                                             
LP001002             1.0         Urban           Y  
LP001003             1.0         Rural           N  
LP001005             1.0         Urban           Y  
LP001006             1.0         Urban           Y  
LP001008             1.0         Urban           Y

In [52]:
import pandas as pd
import numpy as np
from scipy import stats

selected_num_col=df[['ApplicantIncome','CoapplicantIncome','LoanAmount']]

df=df[(np.abs(stats.zscore(selected_num_col)) < 3).all(axis=1)]
df.head()

Gender Married  Dependents     Education Self_Employed  \
Loan_ID                                                           
LP001002   Male      No         0.0      Graduate            No   
LP001003   Male     Yes         1.0      Graduate            No   
LP001005   Male     Yes         0.0      Graduate           Yes   
LP001006   Male     Yes         0.0  Not Graduate            No   
LP001008   Male      No         0.0      Graduate            No   

          ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
Loan_ID                                                                      
LP001002             5849                0.0       126.0             360.0   
LP001003             4583             1508.0       128.0             360.0   
LP001005             3000                0.0        66.0             360.0   
LP001006             2583             2358.0       120.0             360.0   
LP001008             6000                0.0       141.0             360.0   

          Credit_History Property_Area Loan_Status  
Loan_ID                                             
LP001002             1.0         Urban           Y  
LP001003             1.0         Rural           N  
LP001005             1.0         Urban           Y  
LP001006             1.0         Urban           Y  
LP001008             1.0         Urban           Y

In [53]:
cat_col=df.select_dtypes(include=['object'])
cat_col

Gender Married     Education Self_Employed Property_Area Loan_Status
Loan_ID                                                                      
LP001002   Male      No      Graduate            No         Urban           Y
LP001003   Male     Yes      Graduate            No         Rural           N
LP001005   Male     Yes      Graduate           Yes         Urban           Y
LP001006   Male     Yes  Not Graduate            No         Urban           Y
LP001008   Male      No      Graduate            No         Urban           Y
...         ...     ...           ...           ...           ...         ...
LP002971   Male     Yes  Not Graduate           Yes         Urban           Y
LP002975   Male     Yes      Graduate            No         Urban           Y
LP002980   Male      No      Graduate            No     Semiurban           Y
LP002986   Male     Yes      Graduate            No         Rural           N
LP002989   Male      No      Graduate           Yes         Rural           Y

[943 rows x 6 columns]

In [54]:
df=df.dropna()


In [55]:
cat_col.nunique()

Gender           2
Married          2
Education        2
Self_Employed    2
Property_Area    3
Loan_Status      2
dtype: int64

We can see that Gender, Married, Education, Self_Employed, and Loan_Status have 2 values.

In [56]:
# unique values in columns
print(df['Gender'].unique())
print(df['Married'].unique())
print(df['Education'].unique())
print(df['Self_Employed'].unique())
print(df['Loan_Status'].unique())



['Male' 'Female']
['No' 'Yes']
['Graduate' 'Not Graduate']
['No' 'Yes']
['Y' 'N']


In [57]:
df = pd.get_dummies(df, columns=['Property_Area'])
print(df)


         Gender Married  Dependents     Education Self_Employed  \
Loan_ID                                                           
LP001002   Male      No         0.0      Graduate            No   
LP001003   Male     Yes         1.0      Graduate            No   
LP001005   Male     Yes         0.0      Graduate           Yes   
LP001006   Male     Yes         0.0  Not Graduate            No   
LP001008   Male      No         0.0      Graduate            No   
...         ...     ...         ...           ...           ...   
LP002971   Male     Yes         3.0  Not Graduate           Yes   
LP002975   Male     Yes         0.0      Graduate            No   
LP002980   Male      No         0.0      Graduate            No   
LP002986   Male     Yes         0.0      Graduate            No   
LP002989   Male      No         0.0      Graduate           Yes   

          ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
Loan_ID                                           

In [58]:
#Gender, Married, Education, Self_Employed, Credit_History and Loan_Status have 2 values
# replacing values
df['Gender'].replace(['Male', 'Female'],
                        [0, 1], inplace=True)

df['Married'].replace(['No', 'Yes'],
                        [0, 1], inplace=True)
df['Education'].replace(['Graduate', 'Not Graduate'],
                        [0, 1], inplace=True)
df['Self_Employed'].replace(['No', 'Yes'],
                        [0, 1], inplace=True)

df['Loan_Status'].replace(['N', 'Y'],
                        [0, 1], inplace=True)


df.head()

Gender  Married  Dependents  Education  Self_Employed  \
Loan_ID                                                           
LP001002       0        0         0.0          0              0   
LP001003       0        1         1.0          0              0   
LP001005       0        1         0.0          0              1   
LP001006       0        1         0.0          1              0   
LP001008       0        0         0.0          0              0   

          ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
Loan_ID                                                                      
LP001002             5849                0.0       126.0             360.0   
LP001003             4583             1508.0       128.0             360.0   
LP001005             3000                0.0        66.0             360.0   
LP001006             2583             2358.0       120.0             360.0   
LP001008             6000                0.0       141.0             360.0   

          Credit_History  Loan_Status  Property_Area_Rural  \
Loan_ID                                                      
LP001002             1.0            1                    0   
LP001003             1.0            0                    1   
LP001005             1.0            1                    0   
LP001006             1.0            1                    0   
LP001008             1.0            1                    0   

          Property_Area_Semiurban  Property_Area_Urban  
Loan_ID                                                 
LP001002                        0                    1  
LP001003                        0                    0  
LP001005                        0                    1  
LP001006                        0                    1  
LP001008                        0                    1

In [59]:

from sklearn.preprocessing import StandardScaler
import numpy as np

# scaler
scaler = StandardScaler()




In [60]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression



X = df.drop('Loan_Status', axis=1)
y = df['Loan_Status']

X = scaler.fit_transform(X)


# split the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)



# define the model
model = LogisticRegression()

# fit on the training set
model.fit(X_train, y_train)

# predict on test set
y_pred = model.predict(X_test)



print('score:', model.score(X_test, y_test))

print(y_pred)



score: 0.8888888888888888
[1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0 1 1 1
 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 1 0 1 1 0 1 1 1
 1 0 1 1 1 0 1 1 0 0 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 0 1 0 1 0 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 0 1 1 1 1 1
 1 1]


In [61]:
from sklearn import metrics

cnf_matrix = metrics.confusion_matrix(y_test, y_pred)
cnf_matrix

array([[ 29,  26],
       [  3, 203]])

#df['Gender'].replace(['Male', 'Female'],
                        [0, 1], inplace=True)

#df['Married'].replace(['No', 'Yes'],
                        [0, 1], inplace=True)
#df['Education'].replace(['Graduate', 'Not Graduate'],
                        [0, 1], inplace=True)
#df['Self_Employed'].replace(['No', 'Yes'],
                        [0, 1], inplace=True)

#df['Loan_Status'].replace(['N', 'Y'],
                        [0, 1], inplace=True)


Gender                   Male:0
Married                    No:0
Dependents                  0:0
Education            Graduate:0
Self_Employed              No:0
ApplicantIncome          2400
CoapplicantIncome        2000
LoanAmount                 36
Loan_Amount_Term          360
Credit_History              1:1
Property_Area           Urban:0,0,1(dummies brought new columns)

Loan_Status:Y=1 

In [62]:
df.columns

Index(['Gender', 'Married', 'Dependents', 'Education', 'Self_Employed',
       'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Loan_Status',
       'Property_Area_Rural', 'Property_Area_Semiurban',
       'Property_Area_Urban'],
      dtype='object')

In [63]:
#the second number in Loan_Status tells the 'yes'
import warnings
from sklearn.linear_model import LogisticRegression
warnings.filterwarnings("ignore")

print(model.predict_proba(scaler.transform([[0,0,0,0,0,2400, 2000, 36, 360,1,0,0,1]]))[:,1])

[0.87215109]


87% probability the Loan_Status is Yes.

## Handin your final answers by running the code cell below.
- Save your latest changes.
- Please remove anything that may identify you to your anonymous reviewer.
- You may run the code cell as many times as you wish.
- Your permission to write the handin file ends at the deadline.

In [ ]:
import sys
sys.path.append('/home/varpha/data_analytics/lib')
from handin import handin_exrc_06
handin_exrc_06()


All the relevant .ipynb files in your current directory:

  1. 08_04_exrc_06-Copy1.ipynb
  2. NEW_08_04_exrc_06.ipynb

